internal/did-x509-resolver/resolver.go

```diff
@@ -0,0 +1,448 @@
+				case "L":
+					fieldValues = leaf.Subject.Locality
+				case "S":
+					fieldValues = leaf.Subject.Province
+				case "STREET":
+					fieldValues = leaf.Subject.StreetAddress
+				case "POSTALCODE":
+					fieldValues = leaf.Subject.PostalCode
+				case "SERIALNUMBER":
+					fieldValues = []string{leaf.Subject.SerialNumber}
+				case "CN":
+					fieldValues = []string{leaf.Subject.CommonName}
+				default:
+					for _, aav := range leaf.Subject.Names {
+						if aav.Type.String() == k {
+							fieldValues = []string{aav.Value.(string)}
+							break
+						}
+					}
+					if len(fieldValues) == 0 {
+						return fmt.Errorf("unsupported subject key: %s", k)
+					}
+				}
+				found := false
+				for _, fv := range fieldValues {
+					if strings.Contains(fv, v) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return fmt.Errorf("invalid subject value: %s=%s", k, v)
+				}
+			}
+		case "san":
+			if len(args) != 2 {
+				return fmt.Errorf("exactly one SAN type and value required")
+			}
+			sanType := args[0]
+			sanValue, err := url.QueryUnescape(args[1])
+			if err != nil {
+				return fmt.Errorf("url.QueryUnescape failed: %w", err)
+			}
+			err = checkHasSan(sanType, sanValue, chain[0])
+			if err != nil {
+				return err
+			}
+		case "eku":
+			if len(args) != 1 {
+				return errors.New("exactly one EKU required")
+			}
+
+			ekuOid, err := oidFromString(args[0])
+			if err != nil {
+				return fmt.Errorf("oidFromString failed: %w", err)
+			}
+
+			if len(chain[0].UnknownExtKeyUsage) == 0 {
+				return errors.New("no EKU extension in certificate")
+			}
+
+			foundEku := false
+			for _, certEku := range chain[0].ExtKeyUsage {
+				certEkuOid, ok := OidFromExtKeyUsage(certEku)
+				if ok && certEkuOid.Equal(*ekuOid) {
+					foundEku = true
+					break
+				}
+			}
+			for _, certEkuOid := range chain[0].UnknownExtKeyUsage {
+				if certEkuOid.Equal(*ekuOid) {
+					foundEku = true
+					break
+				}
+			}
+
+			if !foundEku {
+				return fmt.Errorf("EKU not found: %s", ekuOid)
+			}
+		case "fulcio-issuer":
+			if len(args) != 1 {
+				return errors.New("excessive arguments to fulcio-issuer")
+			}
+			decodedArg, err := url.QueryUnescape(args[0])
+			if err != nil {
+				return fmt.Errorf("urlUnescape failed: %w", err)
+			}
+			fulcioIssuer := "https://" + decodedArg
+			fulcioIssuerOid, err := oidFromString("1.3.6.1.4.1.57264.1.1")
+			if err != nil {
+				return fmt.Errorf("oidFromString failed: %w", err)
+			}
+			found := false
+			for _, ext := range chain[0].Extensions {
+				if ext.Id.Equal(*fulcioIssuerOid) {
+					if string(ext.Value) == fulcioIssuer {
+						found = true
+						break
+					}
+				}
+			}
+			if !found {
+				return fmt.Errorf("invalid fulcio-issuer: %s", fulcioIssuer)
+			}
+		default:
+			return fmt.Errorf("unsupported did:x509 policy name '%s'", policyName)
+		}
+	}
+
+	return nil
+}
+
+func createDidDocument(did string, chain []*x509.Certificate) (string, error) {
+	format := `
+{
```

Comment:
Feel free but I don't have the confidence that it will be completely correct.

Reply:
Neither do I; I've never seen this kind of json serialization definition before.

Reply:
The one fear would be strings that could contain invalid characters being substituted into the string substitution, but I think we have sufficient control over the contents below to not worry about that for now.  If we're being super paranoid, we could check the `did` string for embedded `"` characters.

Reply:
That's a good point, although to get here, the various strings would have to pass all the other checks that came before. I could imagine someone cooking up an x509 certificate with unusual contents, so it might be worth adding another check for the policy-value (`args`) to be within spec as per https://github.com/microsoft/did-x509/blob/main/specification.md#identifier-syntax .

Reply:
Bryce's point is also one of the reasons I suggested to use a serializer. Here's an example of what I had in mind: https://go.dev/play/p/qzP-tbA0m-U

As you can see, the empty fields are omitted as expected and quotes are automatically escaped.
An added bonus is that you can load a raw did doc into a JSON as well, so it goes both ways. Anyways, lets revisit this later.